pykern/pkunit.py

```diff
@@ -511,6 +391,144 @@ def work_dir():
     return _base_dir(_WORK_DIR_SUFFIX).ensure(dir=True)
 
 
+class _FileEq:
+    """Implements `file_eq`"""
+    def __init__(self, expect_path, *args, **kwargs):
+        self._validate_and_set_paths(expect_path, *args, **kwargs)
+        self._set_expect_and_actual()
+        self._compare()
+
+    def _set_expect_and_actual(self):
+        if self._convert_xlsx():
+            return
+        if self.a:
+            self.actual = self.kwargs['actual']
+            if self.actual_path.exists():
+                pkfail('actual={} and actual_path={} both exist', self.actual, self.actual_path)
+        else:
+            self.actual = pkio.read_text(self.actual_path)
+        if self._set_json():
+            return
+        if self._set_jinja():
+            return
+        self._set_default()
+
+    def _set_json(self):
+        if self.expect_path.ext == '.json' and not self.actual_path.exists():
+            self.expect = pkio.read_text(self.expect_path)
+            if self.a: # TODO (gurhar1133): rename self.a
+                import pykern.pkjson
+                pkio.mkdir_parent_only(self.actual_path)
+                self.actual = pykern.pkjson.dump_pretty(self.actual, filename=self.actual_path)
+            return True
+        return False
+
+    def _set_jinja(self):
+        if self.is_jinja:
+            import pykern.pkjinja
+
+            self.expect = pykern.pkjinja.render_file(self.expect_path, self.kwargs['j2_ctx'], strict_undefined=True)
+            if self.a:
+                pkio.write_text(self.actual_path, self.actual)
+            return True
+        return False
+
+    def _set_default(self):
+        self.expect = pkio.read_text(self.expect_path)
+        if self.a:
+            pkio.write_text(self.actual_path, self.actual)
+
+    def _convert_xlsx(self):
+        if self.expect_path.ext == '.csv' and not self.actual_path.check(file=True):
+            return self._xlsx_to_csv()
+        return False
+
+    def _compare(self):
+        if self.expect == self.actual:
+            return
+        c = f"diff '{self.expect_path}' '{self.actual_path}'"
+        x = self._get_message()
+        with os.popen(c) as f:
+            pkfail(
+                '{}',
+                f'''expect != actual:
+    {c}
+    {f.read()}
+    {x}
```

Comment:
don't need temporary variable now that better modularlization. `{self._message()}`